server/src/main/java/org/opensearch/cluster/routing/WRRShardsCache.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.ClusterChangedEvent;
+import org.opensearch.cluster.ClusterStateListener;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.cache.Cache;
+import org.opensearch.common.cache.CacheBuilder;
+import org.opensearch.common.lease.Releasable;
+import org.opensearch.index.shard.ShardId;
+
+import java.util.List;
+
+/**
+ * The WRR shards cache allows caching shard ordering returned by Weighted round-robin scheduling policy ,helping with
+ * improving similar requests.
+ *
+ * @opensearch.internal
+ */
+public class WRRShardsCache implements Releasable, ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(WRRShardsCache.class);
+
+    private final Cache<Key, List<ShardRouting>> cache;
+
+    public WRRShardsCache(ClusterService clusterService) {
+
+        final long sizeInBytes = 2000000;
+        CacheBuilder<Key, List<ShardRouting>> cacheBuilder = CacheBuilder.<Key, List<ShardRouting>>builder()
+            .removalListener(notification -> logger.info("Object" + " {} removed from cache", notification.getKey().shardId))
+            .setMaximumWeight(sizeInBytes);
+        cache = cacheBuilder.build();
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void close() {
+        logger.info("Invalidating WRRShardsCache on close");
+        cache.invalidateAll();
+    }
+
+    public Cache<Key, List<ShardRouting>> getCache() {
+        return cache;
+    }
+
```

Comment:
Let's not expose the `Cache` outside. It might open doors for unwanted mutations

Reply:
Makes sense. Will change this

Reply:
Made changes